Dialogs/EventArgumentsDialog.cpp

```diff
@@ -0,0 +1,108 @@
+      QSpinBox *integer = new QSpinBox(this);
+      integer->setMinimum(0);
+      value = integer;
+      layout->addWidget(value, row, 1);
+    } else if (arg == "string") {
+      QLineEdit *lineEdit = new QLineEdit(this);
+      lineEdit->setText(tr("MyCustomEvent"));
+      value = lineEdit;
+      layout->addWidget(value, row, 1);
+    } else if (arg == "object") {
+      QHBoxLayout *objLayout = new QHBoxLayout();
+      QToolButton *objButton = new QToolButton(this);
+
+      QMenuView *objMenu = new QMenuView(this);
+      TreeSortFilterProxyModel *treeProxy = new TreeSortFilterProxyModel(this);
+      treeProxy->SetFilterType(TreeNode::TypeCase::kObject);
+      treeProxy->setSourceModel(MainWindow::treeModel.get());
+      objMenu->setModel(treeProxy);
+      objButton->setMenu(objMenu);
+      objButton->setPopupMode(QToolButton::MenuButtonPopup);
+
+      QLineEdit *lineEdit = new QLineEdit(this);
+      lineEdit->setReadOnly(true);
+      QModelIndex firstObjIdx = treeProxy
+                                    ->match(treeProxy->index(0, 0), TreeModel::UserRoles::TypeCaseRole,
+                                            TypeCase::kObject, 1, Qt::MatchRecursive)
+                                    .first();
+      QString firstObj = firstObjIdx.data(Qt::DisplayRole).toString();
+      objButton->setIcon(firstObjIdx.data(Qt::DecorationRole).value<QIcon>());
+      lineEdit->setText(firstObj);
+
+      connect(objMenu, &QMenuView::triggered, [=](const QModelIndex &index) {
+        lineEdit->setText(treeProxy->data(index, Qt::DisplayRole).toString());
+        objButton->setIcon(treeProxy->data(index, Qt::DecorationRole).value<QIcon>());
+      });
+
+      objLayout->addWidget(lineEdit);
+      objLayout->addWidget(objButton);
+
+      layout->addItem(objLayout, row, 1);
+
+      value = lineEdit;
+    } else {
+      QComboBox *combo = new QComboBox(this);
+      auto argList = MainWindow::GetEventData()->value_names_for_type(name->text().toStdString());
+      for (auto a : argList) {
+        combo->addItem(QString::fromStdString(a.first));
+      }
+      value = combo;
+      layout->addWidget(value, row, 1);
+    }
+
+    widgets_.append(value);
+
+    layout->addWidget(name, row, 0);
+    row++;
+  }
+
+  QDialogButtonBox *btn = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
+  layout->addWidget(btn, row, 1);
+  connect(btn, SIGNAL(accepted()), this, SLOT(accept()));
+  connect(btn, SIGNAL(rejected()), this, SLOT(reject()));
+
+  setWindowTitle(tr("Event arguments"));
+}
+
+const QStringList &EventArgumentsDialog::GetArguments() const { return arguments_; }
+
+void EventArgumentsDialog::done(int r) {
+  for (const QWidget *w : widgets_) {
+    QVariant argument = w->metaObject()->userProperty().read(w);
+    QString argstr = "";
+    if (QString(w->metaObject()->className()) == "QSpinBox") {
+      argstr = QString::number(reinterpret_cast<const QSpinBox *>(w)->value());
+    } else {
```

Comment:
Why did you do that? It should have worked the way it was, the user property of a spinbox is the value.
`Q_PROPERTY(int value READ value WRITE setValue NOTIFY valueChanged USER true)`
https://code.woboq.org/qt5/qtbase/src/widgets/widgets/qspinbox.h.html#62

Reply:
Clearly I did it because it didn't work. a) your original code was never triggered for spinboxes and b) .toString() changed 33 to "!"

Reply:
I'm going to look at it before I merge, it looks like it should work the way I had it, I'll have to test.